components/scream/src/share/field/field_monotonicity_check.hpp

```diff
@@ -0,0 +1,49 @@
+#ifndef SCREAM_FIELD_MONOTONICITY_CHECK_HPP
+#define SCREAM_FIELD_MONOTONICITY_CHECK_HPP
+
+#include "share/field/field.hpp"
+
+namespace scream
+{
+
+// This field property check returns true if all data in a given field are
+// monotonically increasing or decreasing, and false if not. Currently, no
+// repairs to nonmonotonic fields can be performed. There are ways of repairing
+// non-monotonic fields given certain assumptions, but we do not make such
+// assumptions here.
+template<typename RealType>
+class FieldMonotonicityCheck: public FieldPropertyCheck<RealType> {
+public:
+
+  // Default constructor.
+  FieldMonotonicityCheck () {}
+
+  // Overrides.
+
+  bool check(const Field<RealType>& field) const override {
+    auto view = field.get_view();
+    RealType sign;
+    Kokkos::parallel_reduce(view.extent(0), KOKKOS_LAMBDA(Int i, RealType& s) {
+      if ((i > 0) && (i < view.extent(0)-1)) {
+        auto diff1 = view(i) - view(i-1);
+        auto diff2 = view(i+1) - view(i);
+        s *= (diff1 * diff2 > 0) ? (diff1 * diff2) : 0;
```

Comment:
More of just a question.  Does Kokkos initialize `s` to 1 or something.  My reading of this loop is that we want to multiply by 0 if non-monotonic and so that the product reduce at the end makes the value of sign=0.  But what does `s` start off as?

Reply:
Good question, and an example of how lots of hidden assumptions live within Kokkos. The `Kokkos::Prod<RealType>` operator uses the "multiplicative identity" for RealType, which is 1, so that's what the `sign` variable is initialized to (and `s` is applied to it). Similarly, there's a `Kokkos::Sum<RealType>` that uses the "additive identity", which is 0.

Reply:
Ahhh, very cool and slick.